apps/wallet-api-tools/src/Editor.tsx

```diff
@@ -1,92 +1,121 @@
+    () =>
+      [...history, ...infoHistory].sort(
+        (a, b) => new Date(a.date).getTime() - new Date(b.date).getTime(),
+      ),
+    [history, infoHistory],
+  );
+
   const searchParams = useSearchParams();
-  const codeMirrorRef = useRef<ReactCodeMirrorRef | null>(null);
+  const panelRef = useRef<HTMLDivElement>(null);
   const transportRef = useRef<Transport | null>(null);
+  const [value, setValue] = useState("");
 
   const isSimulator = searchParams.get("simulator");
 
   const pushMessage = useCallback(
     (type: Message["type"], value: Message["value"]) => {
-      setHistory((prev) => [...prev, { date: new Date(), type, value }]);
+      if (
+        type === "error" ||
+        type === "info" ||
+        (type === "in" && !JSON.parse(value).id)
+      ) {
+        setInfoHistory((prev) => [...prev, { date: new Date(), type, value }]);
+      } else if (type === "out") {
+        const parsedValue = JSON.parse(value) as {
+          id: string;
+        };
+        setHistoryAtom((prev) => [
+          ...prev,
+          {
+            type: "group",
+            id: parsedValue.id,
+            date: new Date(),
+            messages: {
+              sent: { date: new Date(), type, value },
+              received: undefined,
+            },
+          },
+        ]);
+      } else {
+        const parsedValue = JSON.parse(value) as {
+          id: string;
+        };
+        setHistoryAtom((prev) => {
+          const updatedHistory = prev.map((item) => {
+            // Ensure the item is of type MessageGrouped before checking its ID
+            if (item.type === "group" && item.id === parsedValue.id) {
+              // Create a new object instead of mutating the existing one
+              return {
+                ...item,
+                messages: {
+                  ...item.messages,
+                  received: { date: new Date(), type, value },
+                },
+              };
+            }
+            return item;
+          });
+
+          return updatedHistory;
+        });
+      }
     },
-    []
+    [],
   );
 
-  const value = useMemo(() => history
-      .map(
-        (message) =>
-          `${getMessageStatus(message)} ${message.date.toLocaleTimeString()}\n${
-            message.value
-          }`
-      )
-      .join("\n\n\n"), [history]);
-
   const handleMessage = useCallback(
     (message: string) => {
       const prettifiedJson = JSON.stringify(JSON.parse(message), null, 2);
       pushMessage("in", prettifiedJson);
     },
-    [pushMessage]
+    [pushMessage],
   );
 
   useEffect(() => {
-    const codeMirror = codeMirrorRef.current;
+    const timer = setTimeout(() => {
+      if (panelRef.current !== null) {
+        panelRef.current.scrollTop = panelRef.current.scrollHeight;
+      }
+    }, 10);
 
-    if (codeMirror?.editor) {
-      codeMirror.editor.scrollTop = codeMirror.editor.scrollHeight;
-    }
+    return () => clearTimeout(timer);
```

Comment:
I remember we talk about stop doing this scrollTop behaviour whenever the history changes
And instead we should show a button if the user is not at the bottom